src/calculateBullsAndCows.js

```diff
@@ -0,0 +1,39 @@
+const correctWordInPlural = (quantity, word) => {
+  return quantity > 1 ? `${word}s` : word;
+}
+
+export const calculateBullsAndCows = (userInput, generatedNumber) => {
+  const res = [];
+  const updatedUserInput = userInput.split('');
+  const updatedGeneratedNumber = generatedNumber.split('');
+
+  if (userInput.length < 4 || userInput.length > 4) {
+    throw new Error('The number should contain 4 digits');
+  }
+
+  if (new Set(userInput).size !== userInput.length) {
+    throw new Error('All digits should be unique');
+  }
+
+  for (let i = 0; i < 4; i++) {
+    if (updatedGeneratedNumber[i] === updatedUserInput[i]) {
+      res.push('bull')
+      continue;
+    }
+
+    if (updatedGeneratedNumber.includes(updatedUserInput[i])) {
+      res.push('cow')
+    }
+  }
+
+  const numberOfBulls = res.filter(value => value === 'bull').length;
```

Comment:
Instead of using `filter` method to calculate the number of bulls and cows, use `reduce` method which would be more efficient for this purpose.